### ISSUE TITLE
Project Loom does not play well with synchronized, so WDYT of adding support for "forbid use of synchronized"?

### ISSUE THREAD

**Issue by vlsi** [policeman-tools/forbidden-apis#213]
In pgjdbc, we replace all `synchronized` with explicit `Lock`: https://github.com/pgjdbc/pgjdbc/pull/2635
I wonder if I could use `forbidden-apis` to avoid more `synchronized` keywords creeping in the future.

**Comment by uschindler** [comment]
Good idea. But this would be an additional "feature" as it requires not to parse for forbidden methods calls, but just look at attributes of methods.
- I have to check how to detect a synchronized block inside bytecode. For method declartion it is just access flag (as far as I remember).
- Add some setting to enable it globally.
- Make sure SuppressForbidden works.

**Comment by vlsi** [comment]
Well, `synchronized(something)` would use [monitorenter](https://docs.oracle.com/javase/specs/jvms/se15/html/jvms-6.html#jvms-6.5.monitorenter) bytecode (e.g. https://asm.ow2.io/javadoc/org/objectweb/asm/Opcodes.html#MONITORENTER )

`synchronized` methods would indeed be coded as method modifiers.

On top of that, it would probably be great to include `Object#wait`, `Object#notify`, and `Object#notifyAll` into the forbidden set as they make sense only in case `synchronized` is used (they should be replaced with `Condition#await`, `Condition#signal`, `Condition#signalAll`).

I've no idea if `synchronized` methods or `monitorenter` could appear in synthetic code, or in the code generated by Kotlin, Scala, or the other compilers.
However, pgjdbc is Java-only for now, so it would be nice to just forbid

**Comment by uschindler** [comment]
Kotlin must also use monitor enter or attributes on methods, unless they use locks internally. Groovy at least uses it.

**Comment by marinier** [comment]
FWIW, the issue with using `synchronized` with virtual threads will be fixed in Java 24: https://openjdk.org/jeps/491

**Comment by prdoyle** [comment]
This seems not to be relevant any longer. Loom works fine with synchronized.

**Comment by vlsi** [comment]
This is relevant for Java 21 though

**Comment by prdoyle** [comment]
Ah right, my bad.

**Comment by prdoyle** [comment]
>     * I have to check how to detect a synchronized block inside bytecode. For method declartion it is just access flag (as far as I remember).

You'd be looking for a `monitorenter` instruction.

**Comment by uschindler** [comment]
> > ```
> > * I have to check how to detect a synchronized block inside bytecode. For method declartion it is just access flag (as far as I remember).
> > ```
> 
> You'd be looking for a `monitorenter` instruction.

Hi,
thats not the only way you can enter `synchronized` block. You also need to check for calls to `synchronized` methods, which implicitely adds `monitorenter`.

I don't think this feature is really important for forbiddenapis. It requires many small extra code changes in the already complex code (detecting lambdas,...) just for a feature which is "broken" in a specific Java version. In general I recommend all people to use virtual threads only in latest LTS Java 25, before that there are too many bugs with it.

Uwe
